README.md

# Libdnet #
----------

Libdnet provides a simplified, portable interface to several low-level networking routines, including network address manipulation, kernel arp(4) cache and route(4) table lookup and manipulation, network firewalling, network interface lookup and manipulation, IP tunneling, and raw IP packet and Ethernet frame transmission.

This is a more or less personal fork I'm committed to maintain and keep up-to-date for educational proposes but any kind of help is greatly appreciated. The list of changes and "planned adjustments" pretty much matches that of upstream.

original: https://github.com/dugsong/libdnet
upstream: https://github.com/boundary/libdnet
nmap fork of libdnet: https://github.com/nmap/nmap/tree/master/libdnet-stripped

## Documentation

(coming soon)

## Build

For now, see INSTALL for various build instructions.

## Downloads

(coming soon)

## Bug report
Please report any bugs or issues here. If your issue occurs only on a specific OS, please mention it in the report.

## License
That's yet to be determined...

## Changes/Plans
----------
    * merged fixes and features from nmap fork of libdnet-stripped
        https://github.com/nmap/nmap/tree/master/libdnet-stripped
    * merged fixes from https://code.google.com/p/libdnet/issues/list
    * MinGW build compatibility
    * MSVC build compatibility (coming soon)
    * npcap compatibility (coming soon)
    * added os_intf_name and pcap_intf_name attributes
    * to intf.h for gaining access to the raw windows interface names
    * adds htonll/ntohll 64-bit endian conversions
    * various small build fixes


debian/changelog
libdumbnet (1.12-4build1) trusty; urgency=medium

  * Rebuild to drop files installed into /usr/share/pyshared.

 -- Matthias Klose <doko@ubuntu.com>  Sun, 23 Feb 2014 13:48:12 +0000

libdumbnet (1.12-4) unstable; urgency=low

  * QA upload.
    - Set Maintainer to Debian QA Group.
  * Convert to dh_python2 (Closes: #616861).
  * debian/rules:
    - Pass --add-missing parameter to automake call (Closes: #713367).

 -- Luca Falavigna <dktrkranz@debian.org>  Sun, 07 Jul 2013 14:20:17 +0200

libdumbnet (1.12-3.1) unstable; urgency=low

  * Non-maintainer upload.
  * Don't ship .la files.

 -- Luk Claes <luk@debian.org>  Fri, 01 Jul 2011 19:43:39 +0200

libdumbnet (1.12-3) unstable; urgency=low

  * Restore lost libdumbnet.so symlink in the -dev package.
    Closes: #542972.
  * Provide dumbnet-config again, provide the old name 'dnet-config'
    as a symlink, and make it print the correct -l parameter (i.e.
    include it into the changed files of the 01rename_library patch).
    Closes: #543050.
  * Write a terse manpage for dumbnet-config.
  * Bump Standards Version to 3.8.3.
    * Add a README.source file.

 -- Jan Christoph Nordholz <hesso@pool.math.tu-berlin.de>  Sun, 23 Aug 2009 00:09:20 +0200

libdumbnet (1.12-2) unstable; urgency=low

  * kfreebsd has <stropts.h>, too, so the ./configure check to decide
    between tun-solaris and tun-bsd is bogus. For the time being patch
    ./configure.in to drop that test and choose tun-bsd right away.

 -- Jan Christoph Nordholz <hesso@pool.math.tu-berlin.de>  Wed, 29 Jul 2009 13:56:43 +0200

libdumbnet (1.12-1) unstable; urgency=low

  * New upstream version.
  * Update upstream location in all Debian files, but do not overwrite
    the upstream README - add a note to README.Debian instead.
  * Add a symbols file.
  * Bump Standards Version to 3.8.2.

 -- Jan Christoph Nordholz <hesso@pool.math.tu-berlin.de>  Sat, 25 Jul 2009 13:51:29 +0200

libdumbnet (1.11-1) UNRELEASED; urgency=low

  * Adopt the package. Closes: #474128.
  * New upstream version.
  * Completely rework the packaging. Use dpatch for patch management.
  * Thoroughly revise debian/copyright.
  * Accomplish the Debian-specific library renaming by modifying the
    .am files and rerunning the whole auto-toolchain. This is the cleanest
    way to do this and produces the smallest distro diff.
    -> New dpatch (script): 01rename_library.sh
  * Activate the python module build (closes: #388541):
    * Configure --with-python.
    * Patch the .pyx code to avoid using the reserved keyword 'type' as a
      mundane identifier.
      -> New dpatch: 02python_type_is_reserved
    * Patch the python Makefile to build against all current python versions.
      -> New dpatch: 03build_all_python_versions
  * Use debhelper v7.
  * Standards Version is now 3.8.1.

 -- Jan Christoph Nordholz <hesso@pool.math.tu-berlin.de>  Sat, 28 Mar 2009 01:51:45 +0100

libdumbnet (1.8-2) unstable; urgency=low

  * QA upload.
    + Set maintainer to Debian QA Group <packages@qa.debian.org>.
  * Move {Source-Version} to {binary:Version}.
  * Change pwd to CURDIR in rules.
  * Add {shlibs:Depends} to dev package.
  * Add watch file.
  * Remove redundant section: libs from binary package.
  * Bump debhelper build-dep and compat to 5.
  * Bump Standards Version to 3.8.0.

 -- Barry deFreese <bdefreese@debian.org>  Fri, 23 Jan 2009 18:00:52 -0500

libdumbnet (1.8-1.5) unstable; urgency=low

  * Porter NMU.
  * Added support for GNU/kFreeBSD, patch from the BTS (closes: bug#341598).

 -- Aurelien Jarno <aurel32@debian.org>  Sun, 10 Dec 2006 20:38:30 +0100

libdumbnet (1.8-1.4) unstable; urgency=low

  * NMU
  * Fix a bug in addr_btos() (Closes: #364821)

 -- Julien Danjou <acid@debian.org>  Tue, 16 May 2006 13:56:32 +0200

libdumbnet (1.8-1.3) unstable; urgency=low

  * Fixed SONAME problem by fixing the -version-info number in
    src/Makefile.am (Closes: #289777)

 -- Javier Fernandez-Sanguino Pen~a <jfs@computer.org>  Tue, 11 Jan 2005 16:13:41 +0100

libdumbnet (1.8-1.2) unstable; urgency=low

  * Introduce AM_MAINTAINER_MODE in configure.in to avoid timestamp
    skew of the auto* files, also touch ac files in debian/rules configure
    just in case (Closes: #289794)
  * Check if symlinks exist before creating them (in config.status
    and distclean target in debian/rules)

 -- Javier Fernandez-Sanguino Pen~a <jfs@computer.org>  Tue, 11 Jan 2005 14:55:00 +0100

libdumbnet (1.8-1.1) unstable; urgency=low

  * NMU with maintainer's approval.
  * New upstream release (Closes: #253572)
   - Updated config.sub and config.guess from automake 1.7
   - Fixed configure.in since there is no need now to normalize LTLIBOBJS
   - Rerun aclocal, automake and autoconf
   - Updated debian/patches/libdumbnet.patch
   - Change debian/rules so that it will only distclean if the Makefile
     is there and so that it will create the symlinks it needs for cleaning
     if they are not available (will happen if the 'Makefile distclean' breaks
     at some point)

 -- Javier Fernandez-Sanguino Pen~a <jfs@computer.org>  Thu, 23 Dec 2004 11:42:14 +0100

libdumbnet (1.7-3) unstable; urgency=low

  * Fix acinclude.m4, aclocal.m4, configure.in, and configure to test
    correctly for PF_PACKET support on GNU/Linux.  This will cause
    libdumbnet to include the right eth-$ARCH.o file.  (Closes: Bug#219303)

 -- Simon Law <sfllaw@debian.org>  Thu, 13 Nov 2003 01:44:56 -0500

libdumbnet (1.7-2) unstable; urgency=low

  * Use netinet/* headers instead of linux/* headers in src/fw-ipchains.c

 -- Simon Law <sfllaw@debian.org>  Thu, 09 Oct 2003 09:52:00 -0400

libdumbnet (1.7-1) unstable; urgency=low

  * New upstream release.  (Closes: Bug#193777)
  * Added a src/net/sock.h since its contents are unneeded.
    (Closes: Bug#207698)
  * Removed the period from the short description.
  * Moved libdumbnet-dev to libdevel.
  * Makefile.am: Set SONAME to 1.0.0.

 -- Simon Law <sfllaw@debian.org>  Tue,  7 Oct 2003 17:52:59 -0400


src/fw-pf.c
/*
 * fw-pf.c
 *
 * Copyright (c) 2001 Dug Song <dugsong@monkey.org>
 *
 * $Id: fw-pf.c 608 2005-03-15 05:05:37Z dugsong $
 */

#include "config.h"

#include <sys/types.h>
#include <sys/ioctl.h>
#include <sys/socket.h>

#include <net/if.h>
#include <netinet/in.h>
#include <net/pfvar.h>

#include <assert.h>
#include <errno.h>
#include <fcntl.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>

#include "dumbnet.h"

/*
 * XXX - cope with moving pf API
 */
#if defined(DIOCRCLRTABLES)
/* XXX - can't isolate the following change:
 *     $OpenBSD: pfvar.h,v 1.112 2002/12/17 12:30:13 mcbride Exp $
 *  so i'll take 1.119's DIOCRCLRTABLES - 12 days of pf unsupported.
 */
# define HAVE_PF_CHANGE_GET_TICKET	1
/* OpenBSD 3.3+ - 3.6 */
/*     $OpenBSD: pfvar.h,v 1.197 2004/06/14 20:53:27 cedric Exp $ */
/*     $OpenBSD: pfvar.h,v 1.130 2003/01/09 10:40:45 cedric Exp $ */
/*     $OpenBSD: pfvar.h,v 1.127 2003/01/05 22:14:23 dhartmei Exp $ */
# define PFRA_ADDR(ra)	(ra)->addr.v.a.addr.v4.s_addr
# define PFRA_MASK(ra)	(ra)->addr.v.a.mask.v4.s_addr
# define pfioc_changerule	pfioc_rule
# define oldrule	rule
# define newrule	rule
#elif defined(DIOCBEGINADDRS)
/*     $OpenBSD: pfvar.h,v 1.102 2002/11/23 05:16:58 mcbride Exp $ */
# define PFRA_ADDR(ra)	(ra)->addr.addr.v4.s_addr
# define PFRA_MASK(ra)	(ra)->addr.mask.v4.s_addr
#elif defined(PFRULE_FRAGMENT)
/* OpenBSD 3.2 */
/*     $OpenBSD: pfvar.h,v 1.68 2002/04/24 18:10:25 dhartmei Exp $ */
# define PFRA_ADDR(ra)	(ra)->addr.addr.v4.s_addr
# define PFRA_MASK(ra)	(ra)->mask.v4.s_addr
#elif defined (PF_AEQ)
/* OpenBSD 3.1 */
/*     $OpenBSD: pfvar.h,v 1.51 2001/09/15 03:54:40 frantzen Exp $ */
# define PFRA_ADDR(ra)	(ra)->addr.v4.s_addr
# define PFRA_MASK(ra)	(ra)->mask.v4.s_addr
#else
/* OpenBSD 3.0 */
# define PFRA_ADDR(ra)	(ra)->addr
# define PFRA_ADDR(ra)	(ra)->mask
#endif

struct fw_handle {
	int	fd;
};

static void
fr_to_pr(const struct fw_rule *fr, struct pf_rule *pr)
{
	memset(pr, 0, sizeof(*pr));

	strlcpy(pr->ifname, fr->fw_device, sizeof(pr->ifname));

	pr->action = (fr->fw_op == FW_OP_ALLOW) ? PF_PASS : PF_DROP;
	pr->direction = (fr->fw_dir == FW_DIR_IN) ? PF_IN : PF_OUT;
	pr->proto = fr->fw_proto;

	pr->af = AF_INET;
	PFRA_ADDR(&pr->src) = fr->fw_src.addr_ip;
	addr_btom(fr->fw_src.addr_bits, &(PFRA_MASK(&pr->src)), IP_ADDR_LEN);

	PFRA_ADDR(&pr->dst) = fr->fw_dst.addr_ip;
	addr_btom(fr->fw_dst.addr_bits, &(PFRA_MASK(&pr->dst)), IP_ADDR_LEN);

	switch (fr->fw_proto) {
	case IP_PROTO_ICMP:
		if (fr->fw_sport[1])
			pr->type = (u_char)(fr->fw_sport[0] &
			    fr->fw_sport[1]) + 1;
		if (fr->fw_dport[1])
			pr->code = (u_char)(fr->fw_dport[0] &
			    fr->fw_dport[1]) + 1;
		break;
	case IP_PROTO_TCP:
	case IP_PROTO_UDP:
		pr->src.port[0] = htons(fr->fw_sport[0]);
		pr->src.port[1] = htons(fr->fw_sport[1]);
		if (pr->src.port[0] == pr->src.port[1]) {
			pr->src.port_op = PF_OP_EQ;
		} else
			pr->src.port_op = PF_OP_IRG;

		pr->dst.port[0] = htons(fr->fw_dport[0]);
		pr->dst.port[1] = htons(fr->fw_dport[1]);
		if (pr->dst.port[0] == pr->dst.port[1]) {
			pr->dst.port_op = PF_OP_EQ;
		} else
			pr->dst.port_op = PF_OP_IRG;
		break;
	}
}

static int
pr_to_fr(const struct pf_rule *pr, struct fw_rule *fr)
{
	memset(fr, 0, sizeof(*fr));

	strlcpy(fr->fw_device, pr->ifname, sizeof(fr->fw_device));

	if (pr->action == PF_DROP)
		fr->fw_op = FW_OP_BLOCK;
	else if (pr->action == PF_PASS)
		fr->fw_op = FW_OP_ALLOW;
	else
		return (-1);

	fr->fw_dir = pr->direction == PF_IN ? FW_DIR_IN : FW_DIR_OUT;
	fr->fw_proto = pr->proto;

	if (pr->af != AF_INET)
		return (-1);

	fr->fw_src.addr_type = ADDR_TYPE_IP;
	addr_mtob(&(PFRA_MASK(&pr->src)), IP_ADDR_LEN, &fr->fw_src.addr_bits);
	fr->fw_src.addr_ip = PFRA_ADDR(&pr->src);

 	fr->fw_dst.addr_type = ADDR_TYPE_IP;
	addr_mtob(&(PFRA_MASK(&pr->dst)), IP_ADDR_LEN, &fr->fw_dst.addr_bits);
	fr->fw_dst.addr_ip = PFRA_ADDR(&pr->dst);

	switch (fr->fw_proto) {
	case IP_PROTO_ICMP:
		if (pr->type) {
			fr->fw_sport[0] = pr->type - 1;
			fr->fw_sport[1] = 0xff;
		}
		if (pr->code) {
			fr->fw_dport[0] = pr->code - 1;
			fr->fw_dport[1] = 0xff;
		}
		break;
	case IP_PROTO_TCP:
	case IP_PROTO_UDP:
		fr->fw_sport[0] = ntohs(pr->src.port[0]);
		fr->fw_sport[1] = ntohs(pr->src.port[1]);
		if (pr->src.port_op == PF_OP_EQ)
			fr->fw_sport[1] = fr->fw_sport[0];

		fr->fw_dport[0] = ntohs(pr->dst.port[0]);
		fr->fw_dport[1] = ntohs(pr->dst.port[1]);
		if (pr->dst.port_op == PF_OP_EQ)
			fr->fw_dport[1] = fr->fw_dport[0];
	}
	return (0);
}

#ifdef HAVE_PF_CHANGE_GET_TICKET
static int
_fw_cmp(const struct fw_rule *a, const struct fw_rule *b)
{
	if (strcmp(a->fw_device, b->fw_device) != 0 || a->fw_op != b->fw_op ||
	                a->fw_dir != b->fw_dir || a->fw_proto != b->fw_proto ||
	    addr_cmp(&a->fw_src, &b->fw_src) != 0 ||
	    addr_cmp(&a->fw_dst, &b->fw_dst) != 0 ||
	    memcmp(a->fw_sport, b->fw_sport, sizeof(a->fw_sport)) != 0 ||
	    memcmp(a->fw_dport, b->fw_dport, sizeof(a->fw_dport)) != 0)
		return (-1);
	return (0);
}
#endif

fw_t *
fw_open(void)
{
	fw_t *fw;

	if ((fw = calloc(1, sizeof(*fw))) != NULL) {
		if ((fw->fd = open("/dev/pf", O_RDWR)) < 0)
			return (fw_close(fw));
	}
	return (fw);
}

int
fw_add(fw_t *fw, const struct fw_rule *rule)
{
	struct pfioc_changerule pcr;

	assert(fw != NULL && rule != NULL);
	memset(&pcr, 0, sizeof(pcr));
#ifdef HAVE_PF_CHANGE_GET_TICKET
	{
		struct fw_rule fr;

		if (ioctl(fw->fd, DIOCGETRULES, &pcr) < 0)
			return (-1);
		while ((int)--pcr.nr >= 0) {
			if (ioctl(fw->fd, DIOCGETRULE, &pcr) == 0 &&
			    pr_to_fr(&pcr.rule, &fr) == 0) {
				if (_fw_cmp(rule, &fr) == 0) {
					errno = EEXIST;
					return (-1);
				}
			}
		}
	}
#endif
#ifdef DIOCBEGINADDRS
	{
		struct pfioc_pooladdr ppa;

		if (ioctl(fw->fd, DIOCBEGINADDRS, &ppa) < 0)
			return (-1);
		pcr.pool_ticket = ppa.ticket;
	}
#endif
	pcr.action = PF_CHANGE_ADD_TAIL;
	fr_to_pr(rule, &pcr.newrule);

	return (ioctl(fw->fd, DIOCCHANGERULE, &pcr));
}

int
fw_delete(fw_t *fw, const struct fw_rule *rule)
{
	struct pfioc_changerule pcr;

	assert(fw != NULL && rule != NULL);
	memset(&pcr, 0, sizeof(pcr));
#ifdef HAVE_PF_CHANGE_GET_TICKET
	{
		struct fw_rule fr;
		int found = 0;

		if (ioctl(fw->fd, DIOCGETRULES, &pcr) < 0)
			return (-1);
		while ((int)--pcr.nr >= 0) {
			if (ioctl(fw->fd, DIOCGETRULE, &pcr) == 0 &&
			    pr_to_fr(&pcr.rule, &fr) == 0) {
				if (_fw_cmp(rule, &fr) == 0) {
					found = 1;
					break;
				}
			}
		}
		if (!found) {
			errno = ENOENT;
			return (-1);
		}
	}
#endif
#ifdef DIOCBEGINADDRS
	{
		struct pfioc_pooladdr ppa;

		if (ioctl(fw->fd, DIOCBEGINADDRS, &ppa) < 0)
			return (-1);
		pcr.pool_ticket = ppa.ticket;
	}
#endif
	pcr.action = PF_CHANGE_REMOVE;
	fr_to_pr(rule, &pcr.oldrule);

	return (ioctl(fw->fd, DIOCCHANGERULE, &pcr));
}

int
fw_loop(fw_t *fw, fw_handler callback, void *arg)
{
	struct pfioc_rule pr;
	struct fw_rule fr;
	uint32_t n, max;
	int ret = 0;

	memset(&pr, 0, sizeof(pr));
	if (ioctl(fw->fd, DIOCGETRULES, &pr) < 0)
		return (-1);

	for (n = 0, max = pr.nr; n < max; n++) {
		pr.nr = n;

		if ((ret = ioctl(fw->fd, DIOCGETRULE, &pr)) < 0)
			break;
#ifdef PF_TABLE_NAME_SIZE
		/* XXX - actually in r1.125, not 1.126 */
		if (pr.rule.src.addr.type == PF_ADDR_TABLE ||
		    pr.rule.dst.addr.type == PF_ADDR_TABLE)
			continue;
#endif
		if (pr_to_fr(&pr.rule, &fr) < 0)
			continue;
		if ((ret = callback(&fr, arg)) != 0)
			break;
	}
	return (ret);
}

fw_t *
fw_close(fw_t *fw)
{
	if (fw != NULL) {
		if (fw->fd >= 0)
			close(fw->fd);
		free(fw);
	}
	return (NULL);
}
